sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/stream/DirectStreamObserver.java

```diff
@@ -34,21 +38,66 @@
  */
 @ThreadSafe
 public final class DirectStreamObserver<T> implements StreamObserver<T> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(DirectStreamObserver.class);
+  private static final int DEFAULT_MAX_MESSAGES_BEFORE_CHECK = 100;
+
   private final Phaser phaser;
   private final CallStreamObserver<T> outboundObserver;
+  private final int maxMessagesBeforeCheck;
+
+  private int numberOfMessagesBeforeReadyCheck;
 
   public DirectStreamObserver(
       Phaser phaser,
       CallStreamObserver<T> outboundObserver) {
+    this(phaser, outboundObserver, DEFAULT_MAX_MESSAGES_BEFORE_CHECK);
+  }
+
+  DirectStreamObserver(
+      Phaser phaser,
+      CallStreamObserver<T> outboundObserver,
+      int maxMessagesBeforeCheck) {
     this.phaser = phaser;
     this.outboundObserver = outboundObserver;
+    this.maxMessagesBeforeCheck = maxMessagesBeforeCheck;
   }
 
   @Override
   public void onNext(T value) {
-    int phase = phaser.getPhase();
-    if (!outboundObserver.isReady()) {
-      phaser.awaitAdvance(phase);
+    numberOfMessagesBeforeReadyCheck += 1;
+    if (numberOfMessagesBeforeReadyCheck >= maxMessagesBeforeCheck) {
+      numberOfMessagesBeforeReadyCheck = 0;
+      int waitTime = 1;
+      int totalTimeWaited = 0;
+      int phase = phaser.getPhase();
+      while (!outboundObserver.isReady()) {
+        try {
+          phaser.awaitAdvanceInterruptibly(phase, waitTime, TimeUnit.SECONDS);
+        } catch (TimeoutException e) {
+          totalTimeWaited += waitTime;
+          waitTime = waitTime * 2;
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+          throw new RuntimeException(e);
+        }
+      }
+      if (totalTimeWaited > 0) {
+        // If the phase didn't change, this means that the installed onReady callback had not
+        // been invoked.
+        if (phase == phaser.getPhase()) {
+          LOGGER.info(
+              "Output channel stalled for {}s, outbound thread {}. gRPC requires that outbound "
```

Comment:
I'm not 100% why the requirements are being documented in this log.

Reply:
Its to help others in the future with debugging. It took a while to figure out what the issue was.

Reply:
Updated the comment and referred to the JIRA.